Sharding metrics per node via fieldSelector (#1864)
kubernetes/kube-state-metrics: README.md

The downside of using an auto-sharded setup comes from the rollout strategy supported by `StatefulSet`s. When managed by a `StatefulSet`, pods are replaced one at a time with each pod first getting terminated and then recreated. Besides such rollouts being slower, they will also lead to short downtime for each shard. If a Prometheus scrape happens during a rollout, it can miss some of the metrics exported by kube-state-metrics.

### Daemonset sharding for pod metrics

For pod metrics, they can be sharded per node with the following flag:
* `--nodename`

Each kube-state-metrics pod uses FieldSelector (spec.nodeName) to watch/list pod metrics only on the same node.

A daemonset kube-state-metrics example:
```
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: kube-state-metrics
spec:
  template:
    spec:
      containers:
      - name: kube-state-metrics
        args:
        - --resources=pods

>         - --nodename=$(NODE_NAME)

Reviewer: Can you include the downward API in here https://kubernetes.io/docs/concepts/workloads/pods/downward-api/ It's not clear where `$NODE_NAME` is coming from otherwise.

Author: Updated. NODE_NAME is  from:
```
        env:
        - name: NODE_NAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: spec.nodeName

```

```

> For other metrics, they can sharded via [Horizontal sharding](#horizontal-sharding).

Reviewer: ```suggestion
Other metrics can be sharded via [Horizontal sharding](#horizontal-sharding).
```

The downside of using an auto-sharded setup comes from the rollout strategy supported by `StatefulSet`s. When managed by a `StatefulSet`, pods are replaced one at a time with each pod first getting terminated and then recreated. Besides such rollouts being slower, they will also lead to short downtime for each shard. If a Prometheus scrape happens during a rollout, it can miss some of the metrics exported by kube-state-metrics.

### Daemonset sharding for pod metrics

For pod metrics, they can be sharded per node with the following flag:
* `--nodename`

Each kube-state-metrics pod uses FieldSelector (spec.nodeName) to watch/list pod metrics only on the same node.

> A daemonset kube-state-metrics example:

Author: Yes. In adding example in https://github.com/kubernetes/kube-state-metrics/pull/1873, which needs a valid [kube-state-metrics image version]( https://github.com/kubernetes/kube-state-metrics/pull/1873/files#diff-ccfe7f5f6d632e90aa1610bbe66063b81c593bd8fff4e04342653f89c078424eR23).
